### ISSUE TITLE
Getting spammed with clipboard error messages

### ISSUE THREAD

**Issue by DarkDefender** [neovim/neovim#7054]
- `nvim --version`: NVIM v0.2.1-576-gc40093f47
- Vim (version: ) behaves differently? NaN
- Operating system/version: Gentoo
- Terminal name/version: urxvt
- `$TERM`: rxvt-unicode-256color

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
:set clipboard=unnamedplus
```

### Actual behaviour
When triggering copy to clipboard actions, like deleting text or yanking, I will from that point on be spammed with messages like:

```
clipboard: error invoking xclip:   Waiting for selection request number 4
clipboard: error invoking xclip:   Waiting for selection request number 5
clipboard: error invoking xclip:   Waiting for selection request number 6
```
And it will never stop it seems. It will just keep going and counting upwards forever. I had to write a some text in a file and it kept going until I quit neovim. IIRC the number then was something like 1648

### Expected behaviour
Do not spam the user with these messages. However I suspect there is something going wrong and that is why these are spammed so I'm up for some testing to see why this is happening. 

But currently I have no idea what is wrong as it seems like most things are copied to clipboard correctly. However I did notice that it failed to copy stuff sometimes but it doesn't happen that often.

**Comment by jamessan** [comment]
The clipboard provider was recently changed to display the provider's stderr via `:echomsg` so that there's a way for users to get error information.  Apparently xclip, even with `-quiet` specified (unless you overwrote the command), dumps non-error data to stderr.

**Comment by jamessan** [comment]
`-silent` forks into the background, unlike `-quiet`.  I'm not sure if that would cause any problems with using xclip, though.  It's worth testing.

**Comment by bfredl** [comment]
Isn't the problem rather that we don't have `-quitet` also for paste?

**Comment by DarkDefender** [comment]
After some testing outside of neovim, it seems like the problem is that neovim uses the `-quiet` flag. Per default xclip is run with the `-silent` flag and will not pollute the terminal with output.

xclip when run in quite mode will output the "waiting for selection" messages until an other application takes over the clipboard. That means that these messages will be spammed to neovim until something else uses the clipboard. And then it will just trigger again when neovim uses the clipboard...

**Comment by jamessan** [comment]
> Isn't the problem rather that we don't have -quitet also for paste?

No, because paste runs and then exits.  Copy leaves xclip running so it can respond to other paste requests until a different application asserts ownership of the selection.

> Per default xclip is run with the -silent flag and will not pollute the terminal with output.

Yes, but that also causes xclip to detach from the terminal.  I don't know if that's what we want, which is why I suggested testing the use of `-silent`.

**Comment by bfredl** [comment]
We certainly do not want it to fork, that will break `clipboard=unnamed`. Better would be to save the stderr message, and only display it if the process exits with nonzero.

**Comment by bfredl** [comment]
in fact, this is what the python provider code does, and it could be a pattern if not builtin, at least a helper function in the runtime. #6844 will also make this simpler.

**Comment by DarkDefender** [comment]
I removed the quiet flag and I can now copy without and error messages. However, I guess this won't help much it it breaks `clipboard=unnamed`

**Comment by jamessan** [comment]
> We certainly do not want it to fork, that will break clipboard=unnamed.

Why would that break?  A new xclip instance would be spawned, causing the old one to exit since something else has asserted ownership.

If the concern is that we wouldn't be able to send new data to the provider, we already don't do that.  The last thing that `clipboard.set` does is close `stdin`.

**Comment by bfredl** [comment]
"break" was maybe a too strong word, but a lot of users complained that register types was broken with `clipboard=unnamed`, before this logic was introduced to fix it.

**Comment by jamessan** [comment]
Which logic?  Being able to re-use the data in `s:selections[a:reg].data` if `s:selects[a:reg].owner` is non-zero?

**Comment by bfredl** [comment]
Indeed, it remembers the register type as well as being a slight performance improvement

**Comment by musaffa** [comment]
I run into this error every time I try to open `netrw` with `:ex`.

**Comment by yzhuang** [comment]
For whoever affected by this issue, I found installing 'xsel' to be a temp workaround (my os is ubuntu 16.04 LTS).

**Comment by khughitt** [comment]
I realize this issue is closed, but just in case anyone else comes across this problem, another possible solution may be to modify your `~/.ssh/config` to enable the `ForwardX11Trusted` flag: 

```
ForwardX11 yes
ForwardX11Trusted yes
```

...Or include the `-Y` flag in our `ssh` command.

**Comment by aikawa9376** [comment]
Will appear again in recent updates
Can other people see it?

build is RelWithDebInfo.
Error message -
clipboard: error invoking xclip: Waiting for selection requests, Control-C to quit   Waiting for selection request number 1   Waiting for selection request number 2

**Comment by petobens** [comment]
I'm also seeing this again.

**Comment by blueyed** [comment]
Can you git-bisect it, please?

**Comment by justinmk** [comment]
@blueyed given https://github.com/neovim/neovim/commit/722806a1154a004494ef911c2b77a2f35a8a4a62 maybe it's related to the new signal handling?

**Comment by Shougo** [comment]
I also get the same error messages.
I can git-bisect later.

**Comment by Shougo** [comment]
I have tested current HEAD.  Same behavior.

**Comment by blueyed** [comment]
@justinmk 
Good call.

@Shougo 
Please test if 939d905 is causing it.

I'd also like to see steps for reproducing this then.
`xclip` is used for me also, but I do not see this issue - but I am also using a clipboard manager on top.

**Comment by Shougo** [comment]
I have bisected.

> Please test if 939d905 is causing it.

Yes.  This is the first bad commit.

> I'd also like to see steps for reproducing this then.

Hm.  Visual yank from bottom lines to top lines is reproducible for me.

**Comment by blueyed** [comment]
@Shougo 
Thanks.
Happens for me actually also after using it twice - apparently I am not pasting much to the system clipboard - or do not pay enough attention.. ;)
See https://github.com/neovim/neovim/pull/10765

**Comment by khughitt** [comment]
Issue is occurring once more for me in Neovim nightly from Jan 5, 2020 :/

- NVIM v0.5.0-296-gfa9b057d3
- xclip 0.13
- `:checkhealth` looks good and picks up xclip

It looks like the regression fix should have been merged into master (and included in the v0.4.0 release..)

Any thoughts on how the issue might still be arising?

**Comment by blueyed** [comment]
@khughitt 
I think it would be helpful if you could debug/investigate based on the code being used there (start by looking at https://github.com/neovim/neovim/pull/10765/files I guess).

**Comment by blueyed** [comment]
@khughitt 
And if it happened only after some update of nvim's nightly, you might want to git-bisect it.

**Comment by khughitt** [comment]
@blueyed Thanks - I will try and narrow down the cause of the issue the next time it arises. Unfortunately, I'm still haven't found a way to consistently reproduce it -- sometimes it occurs very frequently, and other times nothing.

Good suggestion about git-bisect. The issue also appears to occur with the most recent stable version of neovim in the Arch repos (0.4.3), however, so I think it's not clear when the issue was reintroduced (or perhaps it was never resolved for the particular environment I'm using).

I'll post an update if I'm able to figure out anything more informative.

**Comment by yuriescl** [comment]
This is happening on Fedora 34 beta.
```
$ dnf list installed | grep neovim
neovim.x86_64  0.4.4-4.fc34

$ dnf list installed | grep xclip
xclip.x86_64   0.13-12.git9344507.fc34

$ dnf list installed | grep xsel
xsel.x86_64   1.2.0-28.fc34
```

![image](https://user-images.githubusercontent.com/26092447/114302991-11dce200-9aa2-11eb-896b-bf4cbd15945d.png)

**Comment by kting28** [comment]
`a:data > 0 && a:data < 128` fixes it for me.  it's returning `143` now so != check will keep dumping the error message.

**Comment by sisrfeng** [comment]
> I realize this issue is closed, but just in case anyone else comes across this problem, another possible solution may be to modify your `~/.ssh/config` to enable the `ForwardX11Trusted` flag:
> 
> ```
> ForwardX11 yes
> ForwardX11Trusted yes
> ```


Works for me
